File: AD/PromptExercises/mapReduction.py
class Mapper:
    def setup(self):
        # This method is executed once at the beginning of the task.
        # You can perform setup or configuration here.
        pass

    def map(self, key, value):
        # This method implements the logic to process the input data.
        # In this example, we split each line into words and emit a key-value pair for each word.
        for word in value.split():
            self.emit(word, 1)

    def cleanup(self):
        # This method is executed once at the end of the map task.
        # You can close resources or perform any final computations here.
        pass

    def emit(self, key, value):
        # Helper method to emit a key-value pair using the context or collector.
        # The actual implementation may vary based on the MapReduce framework used.
        print(f"{key}, {value}")

# Example usage:
# Create an instance of the Mapper class
mapper = Mapper()

# Call the setup method (usually executed once at the beginning)
mapper.setup()

# Example input data
input_key = 1  # For illustration purposes, the key is not used in this example
input_value = "Hello World Hello MapReduce"

# Call the map method with the input key-value pair
mapper.map(input_key, input_value)

# Call the cleanup method (usually executed once at the end)
mapper.cleanup()

File: AD/PromptExercises/mapReduction.java
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Mapper;

import java.io.IOException;

public class WordCountMapper extends Mapper<LongWritable, Text, Text, IntWritable> {

    @Override
    protected void setup(Context context) throws IOException, InterruptedException {
        // Initialization tasks for the mapper (if needed)
    }

    @Override
    protected void map(LongWritable key, Text value, Context context)
            throws IOException, InterruptedException {
        // Split the line into words
        String[] words = value.toString().split("\\s+");

        // Emit (word, 1) for each word
        for (String word : words) {
            context.write(new Text(word), new IntWritable(1));
        }
    }

    @Override
    protected void cleanup(Context context) throws IOException, InterruptedException {
        // Finalization tasks for the mapper (if needed)
    }
}
